### DISCUSSION TITLE
How to handle .sqbpro files?

### DISCUSSION THREAD

**Discussion by tamununaaa** [sqlitebrowser/sqlitebrowser#3558] [Q&A]
I am working on sqlite database compression. Currently, 32 lakh files are taking 1.41 GB. I want 1 crore files to take <700MB.
I came across a few techniques to compress sqlite DB such as PRAGMA, VACUUM, compression using bzip2 and zip, etc.
PRAGMA and VACUUM are compressing files of size 256MB to 1KB file. But, they are saving the db files as .sqbpro files (sqlite browser project). The issue is that .sqbpro files (of 1KB) cannot open individually, i.e. if I delete the original .sqlite file, then .sqbpro file can open but doesn't show any data.
Is there a way to work with these .sqbpro files? I am okay with using support files.
Or can you help me with any other alternatives?

**Comment by FriedrichFroebel** [comment]
> PRAGMA and VACUUM are compressing files of size 256MB to 1KB file.

You can most likely run these commands through the SQLite3 CLI or with your favorite programming language bindings as well - no need to use DB4S for this.

> But, they are saving the db files as .sqbpro files (sqlite browser project).

This is wrong - as well as the 1 KB file size you claim. `.sqbpro` files are the project files for DB4S and contain a reference to the real database file path as well as some settings, queries etc. If this would really hold all the data directly (and thus provide compression of 256 MB of data to 1 KB of data), this would mean that we have probably found some perfect compression algorithm never seen before ;) To measure the real impact of your `VACUUM` queries and related queries, you would still need to have a look at the actual database file, not at the project file.

**Comment by mxbdev** [comment]
They are XML files and can be easily viewed with Firefox which gives you a clean foldable view unlike other viewers that show one big long string. Unfortunately I dont know where you can find the DTD desription for the XML.

The file just has pointers (like href) to the data or SQL scripts and stores the table schemas in it, but not the data.

I would suggest just zipping or gzipping the *.sqlpro files if you want since they are pure ascii or Unicode characters, but do not delete them afterwards.  The location of the original database is still stored inside.   If you move everything to another directory you will need to unzip its project file, then edit the location pointer.

Better idea is to set up a script to run before compressing (with zip) that will be applied to every .sqlpro file to do the following:
+ extract the database location and print/append it to a file
+ optionally edit in-line the location string and output it.
+ move the database itself from the old location to the new location

You could simply zip the databases too, rather than using some DB tool, but then you must unzip them and the project file to be able to use it.
